swarm/network/stream/delivery.go

```diff
@@ -191,57 +168,84 @@ func (d *Delivery) Close() {
+		// do not send request back to peer who asked us. maybe merge with SkipPeer at some point
+		if req.Origin.String() == id.String() {
+			return true
 		}
-	} else {
-		d.kad.EachConn(req.Addr[:], 255, func(p *network.Peer, po int) bool {
-			id := p.ID()
-			if p.LightNode {
-				// skip light nodes
-				return true
-			}
-			if req.SkipPeer(id.String()) {
-				log.Trace("Delivery.RequestFromPeers: skip peer", "peer id", id)
-				return true
-			}
-			sp = d.getPeer(id)
-			// sp is nil, when we encounter a peer that is not registered for delivery, i.e. doesn't support the `stream` protocol
-			if sp == nil {
-				return true
-			}
-			spID = &id
+
+		// skip peers that we have already tried
+		if req.SkipPeer(id.String()) {
+			log.Trace("findpeer skip peer", "peer", id, "ref", req.Addr.String())
+			return true
+		}
+
+		// if origin is farther away from req.Addr and origin is not in our depth
+		prox := chunk.Proximity(req.Addr, d.kad.BaseAddr())
+		// proximity between the req.Addr and our base addr
+		if po < depth && prox >= depth {
+			log.Trace("findpeer skip peer because depth", "po", po, "depth", depth, "peer", id, "ref", req.Addr.String())
+
+			err = fmt.Errorf("not going outside of depth; ref=%s po=%v depth=%v prox=%v", req.Addr.String(), po, depth, prox)
 			return false
-		})
-		if sp == nil {
-			return nil, nil, errors.New("no peer found")
 		}
+
+		sp = d.getPeer(id)
+
+		// sp is nil, when we encounter a peer that is not registered for delivery, i.e. doesn't support the `stream` protocol
+		return sp == nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if sp == nil {
+		return nil, errors.New("no peer found")
+	}
+
+	return sp, nil
+}
+
+// RequestFromPeers sends a chunk retrieve request to the next found peer
+func (d *Delivery) RequestFromPeers(ctx context.Context, req *storage.Request, localID enode.ID) (*enode.ID, error) {
+	metrics.GetOrRegisterCounter("delivery.requestfrompeers", nil).Inc(1)
+
+	sp, err := d.FindPeer(req)
+	if err != nil {
+		log.Error(err.Error())
+		return nil, err
 	}
 
 	// setting this value in the context creates a new span that can persist across the sendpriority queue and the network roundtrip
 	// this span will finish only when delivery is handled (or times out)
-	ctx = context.WithValue(ctx, tracing.StoreLabelId, "stream.send.request")
-	ctx = context.WithValue(ctx, tracing.StoreLabelMeta, fmt.Sprintf("%v.%v", sp.ID(), req.Addr))
-	log.Trace("request.from.peers", "peer", sp.ID(), "ref", req.Addr)
-	err := sp.SendPriority(ctx, &RetrieveRequestMsg{
+	r := &RetrieveRequestMsg{
 		Addr:      req.Addr,
-		SkipCheck: req.SkipCheck,
-		HopCount:  req.HopCount,
-	}, Top)
+		HopCount:  req.HopCount + 1,
+		SkipCheck: true, // this has something to do with old syncing
+	}
+	log.Trace("sending retrieve request", "ref", r.Addr, "peer", sp.ID().String(), "origin", localID)
+	err = sp.Send(ctx, r)
 	if err != nil {
-		return nil, nil, err
+		log.Error(err.Error())
```

Comment:
Why? `sp.Send` should not error, and we are interested if this happens.

Reply:
yes but i think it is no go style to log a returned error, log it higher up where the error is returned to

Reply:
I think this is an opinion, not really a rule. If you could point me to a blog post, or rule about logging errors in a consistent way, I'd be happy to read it.

In this specific case the reasoning is the following - `RequestFromPeers` can error are two places:
1. Because we have exhausted all peers and can't find a peer.
2. Because we can't send a request to a peer we have found.

1. can happen if we are looking for a chunk that doesn't exist, but it can happen also when we have a bug in our code - so as developers we are interested in logging it, so that we can debug our code.

2. should never really happen, so as developers we are interested in never having an error coming out of `sp.Send` and when this happens, we should understand why it happens.

---

If we don't log different those two cases, how is the caller supposed to know that for us? As far as the caller is concerned, the request for a chunk from peers was not successful, but the underlying reason might be fatal (i.e. network connection issues?) or not (no suitable peer?). As developers (and also users) we are interested in having network connection issues logged as Errors, and no suitable peers maybe not logged as Errors (because then all `chunk not found` cases would also be Errors).

I hope that makes sense.

Reply:
Another point worth considering is the stakeholders involved and consumers of logs... in general I understand them to be:
- regular users of Swarm (they generally have INFO, WARN, ERROR levels enabled)
- developers of Swarm (who can also be users) (they generally have level 4 DEBUG or 5 TRACE enabled)

At the moment the highest priority for logs are to provide value to developers, considering our issues with stability - I'd rather we err on the side of being too verbose (but not spammy), rather than be perfect with our error handling (whatever that means) and risking not providing enough information to developers to debug and solve issues and bugs.

We classify ERRORS as something a developer must look at. In this particular case, if `sp.Send` fails on a deployment, I want to know about it, and see it in the graphs on Grafana, as it really means something is wrong with the peers. Whereas if a `FindPeer` returns an error, I don't want to immediately have to look into it. However I want to be able to look into it, if I know content must be there, and it is not retrievable - hence the TRACE level for its returned error.

Reply:
@zelig relevant - https://dave.cheney.net/2016/06/12/stack-traces-and-the-errors-package

Bottom line - I agree that we should ideally not log an error twice (I think I said that in the previous code review with @janos as well), but I also think it is more important to keep context when we log errors. Since with out current libraries this is not possible, I have resorted to logging errors twice.

Ultimately we need a solution similar to `pkg/errors` and/or a proper logger that has a unique identifier propagated with context (similar to tracing), so that we can correlate traces and ultimately build up the context for an error.

Reply:
Yes, I was suggesting annotating errors to get information about error origins. Package `github.com/pkg/errors` is great and I used it in almost all previous projects, but Go 1.13 is bringing a standard solution `golang.com/x/xerrors` (experiment package) and there are a few presentations/posts about it. Logging and passing errors is something that is a clear problem, buy my opinion, which I explained earlier. I did not read much about it online.

Reply:
I must have missed that suggestion, I only got the part that we shouldn't `handle` (or log) errors and then return them - which technically is correct, but if we don't have a way of keeping the context, I actually disagree with - I think logging is less harmful than not keeping context.

Also annotating manually every single error on every level with `fmt.Errorf` is quite verbose, I am not sure I like this solution.

I will check the `xerrors` package, I wasn't aware of it.

Reply:
What I am trying to say is that annotating errors is the way to keep the context.

Reply:
The question is how we should start annotating errors, because there are multiple ways.

So far I don't think we have done that at many places, and the places we have actually done it is wrong, one place that comes to mind is:
```
func isDecryptError(err error) bool {
	return strings.Contains(err.Error(), api.ErrDecrypt.Error())
}
```

or

```
fmt.Errorf("can't get size: %v", err)
```

This is far from ideal, and conveys less information than trace logs, making it more difficult to debug.

Reply:
Yes, checking substring is not the correct thing to do. I am pretty satisfied with `github.com/pkg/errors` at least with WithMessage and Cause functions. But I would really like to use errors form stdlib from Go 1.13, which can probably happen in a year or so, when we drop the support for Go 1.12. So, would suggest `github.com/pkg/errors` for now.

Reply:
> So far I don't think we have done that at many places, and the places we have actually done it is wrong, one place that comes to mind is:

yeah this is a crappy technical debt on my side. Not a very good example because it's the only one where such ugliness exists :)

Reply:
@justelad I didn't mean to pick on you, actually I remember it was 2-3 of us who added that crap.

Looks like we should add an issue about updating our errors codebase-wide, but I suggest we do that after v0.4 release.

For now I think we can live with such technical debt, considering the rest of the issues we are trying to fix with v0.4, namely (just to name a few):
1. localstore rewrite and migration
2. fetchers rewrite and (hopefully) simplification
3. subscriptions refactor

I agree that logging and returning errors is wrong, but this is the best we have so far in terms of having context and actually detecting issues so that we have "some" way of tracking the underlying issues.

---

Another solution would be to release what we have in `swarm-rather-stable` (i.e. merge to master for the next patch release), and make another round of splitting this PR into smaller chunks, but I'd rather not do that, as it is considerable more effort.